database.go

```diff
@@ -0,0 +1,723 @@
+		ReplicaN: rp.ReplicaN,
+		SplitN:   rp.SplitN,
+	})
+}
+
+// UnmarshalJSON decodes a JSON-encoded byte slice to a retention policy.
+func (rp *RetentionPolicy) UnmarshalJSON(data []byte) error {
+	// Decode into intermediate type.
+	var o retentionPolicyJSON
+	if err := json.Unmarshal(data, &o); err != nil {
+		return err
+	}
+
+	// Copy over properties from intermediate type.
+	rp.Name = o.Name
+	rp.ReplicaN = o.ReplicaN
+	rp.SplitN = o.SplitN
+	rp.Duration = o.Duration
+	rp.Shards = o.Shards
+
+	return nil
+}
+
+// retentionPolicyJSON represents an intermediate struct for JSON marshaling.
+type retentionPolicyJSON struct {
+	Name     string        `json:"name"`
+	ReplicaN uint32        `json:"replicaN,omitempty"`
+	SplitN   uint32        `json:"splitN,omitempty"`
+	Duration time.Duration `json:"duration,omitempty"`
+	Shards   []*Shard      `json:"shards,omitempty"`
+}
+
+// RetentionPolicies represents a list of shard policies.
+type RetentionPolicies []*RetentionPolicy
+
+// Shards returns a list of all shards for all policies.
+func (rps RetentionPolicies) Shards() []*Shard {
+	var shards []*Shard
+	for _, rp := range rps {
+		shards = append(shards, rp.Shards...)
+	}
+	return shards
+}
+
+// Filter represents a tag filter when looking up other tags or measurements.
+type Filter struct {
+	Not   bool
+	Key   string
+	Value string
+	Regex *regexp.Regexp
+}
+
+type Filters []*Filter
+
+// SeriesIDs is a convenience type for sorting, checking equality, and doing union and
+// intersection of collections of series ids.
+type SeriesIDs []uint32
+
+func (p SeriesIDs) Len() int           { return len(p) }
+func (p SeriesIDs) Less(i, j int) bool { return p[i] < p[j] }
+func (p SeriesIDs) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+// Equals assumes that both are sorted. This is by design, no touchy!
+func (a SeriesIDs) Equals(seriesIDs SeriesIDs) bool {
+	if len(a) != len(seriesIDs) {
+		return false
+	}
+	for i, s := range seriesIDs {
+		if a[i] != s {
+			return false
+		}
+	}
+	return true
+}
+
+// Intersect returns a new collection of series ids in sorted order that is the intersection of the two.
+// The two collections must already be sorted.
+func (a SeriesIDs) Intersect(seriesIDs SeriesIDs) SeriesIDs {
+	l := a
+	r := seriesIDs
+
+	// we want to iterate through the shortest one and stop
+	if len(seriesIDs) < len(a) {
+		l = seriesIDs
+		r = a
+	}
+
+	// they're in sorted order so advance the counter as needed.
+	// That is, don't run comparisons against lower values that we've already passed
+	var i, j int
+
+	ids := make([]uint32, 0, len(l))
+	for i < len(l) {
+		if l[i] == r[j] {
+			ids = append(ids, l[i])
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			i += 1
+		} else {
+			j += 1
+		}
+	}
+
+	return SeriesIDs(ids)
+}
+
+// Union returns a new collection of series ids in sorted order that is the union of the two.
+// The two collections must already be sorted.
+func (l SeriesIDs) Union(r SeriesIDs) SeriesIDs {
```

Comment:
because in the above one I set `l` and `r` in the function based on which one has a shorter length. I only reference the receiver when setting those variables.